### ISSUE TITLE
Unit testing for Jupyter Notebooks

### ISSUE THREAD

**Issue by mmcky** [QuantEcon/meta#44]
This looks like a good approach to testing notebooks (without having to put tests inside of the notebook itself). 

When migrating `julia` project over to myst we should explore this option to see if we can have unit testing of the notebooks external to the notebook itself. 

https://github.com/nteract/testbook

cc: @jlperla

**Comment by mmcky** [comment]
It would be nice if `testbook` supported tagged cell blocks that you can use as a reference. Doesn't look like that is a feature yet though.

**Comment by jlperla** [comment]
I don't know... Accessing functions or variables by text in a separate file doesn't sound very appealing, given you convenient it is to have the code next to the test (very different from unit testing of functions/etc. where separate files feels very natural).,   And I don't see how it could work cross-language ?


This sort of stuff is necessary with a raw ipynb since it can't have any skipped cells, but not sure it is needed when there is a text based format which is transformed.

It sure is nice to have the tests next to where the code is used.

The best solution I have found is the `jldoctest`: https://juliadocs.github.io/Documenter.jl/v0.7/man/doctests.html

**Comment by mmcky** [comment]
Thanks for these comments @jlperla -- I am mainly just looking at what is out there to minimise what needs to be developed (+ supported) in EBP. It would be nice if there as a language agnostic Documenter.jl -- I'll take a look to see how that is setup and works.

**Comment by jlperla** [comment]
Yes, thanks for looking out for this. And sorry if I prematurely, and grumpily poopood it.

I think that inside of the notebooks is exactly the right place for it. We don't have that option with jupyter notebooks. But that is part of the reason I think of jupyter as an output format...it isn't a markup language like markdown, myst, rmd, etc. For those, you have lots of flexibility.

The other thing is that jupyterbook doesn't support jupyter as an output format, if I am not mistaken? If not, then it seems like the test infrastructure should be independent of jupyter output (at least until jupyter notebooks are supported). 

One thing that I don't think is necessary (and perhaps not even preferable) is to have the testing code in the actual jupyter or HTML output. It should just be a pass/fail thing during generation itself. In Documenter.jl, etc it always runs jldoctest unless there is a flag set for the compilation. When we hacked the testing for jupinx, we just used the trick to turn them on or off... But I think that was a hack
